torch/library.h

```diff
@@ -298,6 +299,48 @@ namespace detail {
 
   class TorchLibraryInit;
 
+} // namespace detail
+
+// Note [Selective build]
+// ~~~~~~~~~~~~~~~~~~~~~~
+// In some settings, especially mobile, it is important to avoid compiling any
+// references to functions that you aren't actually going to use, so that they
+// can be eliminated by the linker.  We call this capability "selective build".
+//
+// A very easy way to implement selective build which results in a lot of
+// boilerplate is to just add ifdef's around every registration call, but this
+// means you have to write a lot of extra lines of code at every registration
+// site, and it also means you have to define some munging scheme to map
+// operators to macros.
+//
+// Instead of doing this, we have a different mechanism centered around the
+// concept of a SelectiveStr.  A selective name is like a const char* string,
+// except it also carries at compile time a boolean saying whether or not a
+// registration should actually happen or not.  We then have extra overloads which
+// bypass registration entirely if a selective name is disabled.  We do a
+// constexpr test to see if a operator should be enabled or not; this is
+// currently implemented in ATen/core/op_registration/op_whitelist.h
+
+namespace detail {
+
+  // A SelectiveStr is like a const char*, except that it also comes
+  // with a type brand that says whether or not the name is enabled or
+  // not.  If the string is disabled, then (at compile time) we DON'T generate
+  // a registration call for it.  This class is not intended to be called
+  // directly; use TORCH_SELECTIVE_NAME or TORCH_SELECTIVE_SCHEMA macros below
+  // to create it.
+  template <bool enabled>
+  class SelectiveStr {
```

Comment:
Btw, this is an amazing idea and I somehow had forgotten about it. I was trying to find something that avoids having to type the schema string twice at the call site a couple days ago when talking with Martin about selective build for prim ops but didn't come up with this solution :)